### PR TITLE
NET-1475: Implement+test smart contract

### PR DESCRIPTION
**Pull request by Jovonni**
[WIP] 

## Description
Reference implementation for Ceramic Anchor Service V2 smart contract

Include a summary of the change and which issue it addresses in the title of the PR. 

Include relevant motivation, context, brief description and impact of the change(s). List follow-up tasks here.

## TODOs
- CI/CD build pipeline (github actions)
- implement more core logic
- implement more tests
- integration tests for ceramic-anchor-service
- documentation on running end-to-end tests

## How Has This Been Tested?

Describe the tests that you ran to verify your changes. Provide instructions for reproduction.

- [ ] Foundry Tests (solidity) 
    - CeramicAnchorServiceV2.t.sol
       - ceramic-anchor-service/anchor-contract/test/

## Definition of Done
Working implementation of CIP-110 (listed in references)

Before submitting this PR, please make sure:

- [ ] The work addresses the description and outcomes in the issue
- [ ] I have added relevant tests for new or updated functionality
- [ ] My code follows conventions, is well commented, and easy to understand
- [ ] My code builds and tests pass without any errors or warnings
- [ ] I have tagged the relevant reviewers
- [ ] I have updated the READMEs of affected packages
- [ ] I have made corresponding changes to the documentation
- [ ] The changes have been communicated to interested parties

## References:

https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-110/CIP-110.md

### CODE REVIEW

File: anchor-contract/.gitmodules [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[submodule "lib/forge-std"]
+	path = lib/forge-std
+	url = https://github.com/foundry-rs/forge-std
+[submodule "lib/openzeppelin-contracts"]
+	path = lib/openzeppelin-contracts
+	url = https://github.com/openzeppelin/openzeppelin-contracts
```
Comment (oed) [inline anchor-contract/.gitmodules:6]: Can we import these as npm packages instead?

Comment (oed) [reply anchor-contract/.gitmodules:6]: Want to avoid using git submodules.

File: anchor-contract/Makefile [hunk 2]
```diff
@@ -0,0 +1,10 @@
+build:
+	forge build
+t:
+	forge test
+create:
+	forge create CeramicAnchorServiceV2\
+	 --contracts src/CeramicAnchorServiceV2.sol\
+	 --private-key <private-key> --rpc-url https:/infura.io/<key>
+testDeployment:
+	cast call <address> "owner()" --rpc-url https://infura.io/<key>:
```
Comment (oed) [inline anchor-contract/Makefile:10]: Can we include these things in the package.json in the root folder?

**Review by oed — COMMENTED** [review]
Some quick comments on repo structure

Maybe we can rename the `anchor-contract` folder to `contracts`?  That seems more standard.

File: anchor-contract/Makefile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+build:
+	forge build
```
Comment (ukstv) [inline anchor-contract/Makefile:2]: I am curious. Can you install forge via npm?

File: anchor-contract/src/CeramicAnchorServiceV2.sol [hunk 4]
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: GPL
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+struct Service {
```
Comment (ukstv) [inline anchor-contract/src/CeramicAnchorServiceV2.sol:6]: Why struct? Why not just `bool`?

File: anchor-contract/Makefile (hunk 3 above)
Comment (oed) [reply anchor-contract/Makefile:11]: Might be possible through https://www.npmjs.com/package/@foundry-rs/easy-foundryup

**Review by oed — COMMENTED** [review]
Mostly looks good. Just waiting for the move to the pakcage.json instead of Makefile.


Also wondering why do we have two github workflows for foundry tests?
* `anchor-contract/.github/workflows/test.yml`
* `.github/workflows/anchor-contracts.yml`

File: anchor-contract/Makefile (hunk 3 above)
Comment (ukstv) [reply anchor-contract/Makefile:11]: Sweet. Then the whole install/build/test process could be implemented similar to how we do in other packages. @Jovonni

Comment (ukstv) [reply anchor-contract/Makefile:11]: @Jovonni See https://www.twilio.com/blog/npm-scripts and https://docs.npmjs.com/cli/v7/configuring-npm/package-json#config and https://docs.npmjs.com/cli/v7/configuring-npm/npmrc for more information.

File: anchor-contract/test/CeramicAnchorServiceV2.t.sol [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    function testFailAnchorFuzz(bytes calldata _root, address testService) public {
+        vm.assume(testService != address(this)); 
+        casv2.addCas(testService);
+        casv2.anchor(_root);
+    }
+
+    function testCASJs() public {
+        // string[] memory inputs = new string[3];
```
Comment (ukstv) [inline anchor-contract/test/CeramicAnchorServiceV2.t.sol:81]: ?

**Review by ukstv — CHANGES_REQUESTED** [review]
Please, move the install/build/test process from Makefile to package.json, also see Joel's suggestion to use npm-installed forge.

**Review by ukstv — APPROVED** [review]
Submodules suck, but here we are.

🏇

**Review by oed — COMMENTED** [review]
@Jovonni is there no way to make the github action pull down the appropriate submodules? 
Is it perhaps because the submodules are not in the root folder?

Or what's actually the relationship between the submodules and the `contracts/lib` folder?

**Review by oed — APPROVED** [review]
Glad we got that sorted. Lgtm!

**Comment by oed** [comment]
Um, I'm not sure if this is good to merge or not given that the docker image build failed? 
cc @smrz2001

**Comment by smrz2001** [comment]
> > Um, I'm not sure if this is good to merge or not given that the docker image build failed?
> > cc @smrz2001
> 
> Was wondering the same. Here's the failure:
> 
> ''' failed: process "bash --norc -e -o pipefail /bash/scripts/run.sh" did not complete successfully: exit code: 1 4843 3:48PMINFactions.build._exec | #12 794.0 duration=4m15.3s 4844 3:48PMINFactions.build._exec | #12 794.0 3:48PMFTLsystem | failed to execute plan: task failed: actions.verify.healthcheck._exec: process "bash --norc -e -o pipefail /bash/scripts/run.sh" did not complete successfully: exit code: 1 4845 3:48PMINFactions.build._exec | #12 794.0 4846 3:48PMINFactions.build._exec | #12 794.0 4847 3:48PMERRactions.build._exec | failed: process "bash --norc -e -o pipefail /bash/scripts/run.sh" did not complete successfully: exit code: 1 4848 duration=13m14s 4849 3:48PMFTLsystem | failed to execute plan: task failed: actions.build._exec: process "bash --norc -e -o pipefail /bash/scripts/run.sh" did not complete successfully: exit code: 1 '''

I reran the build and it worked, @Jovonni @oed. I'll increase the timeout on the CI healthcheck and continue monitoring for failures.

**Comment by oed** [comment]
Ok, this should be good to merge then!

**Pull request merged**